src/ecs/resources/input.rs

```diff
@@ -62,45 +92,63 @@ impl<'a> Iterator for PressedMouseButtons<'a> {
 /// ```
 #[derive(Default)]
 pub struct InputHandler {
-    pressed_keys: HashMap<VirtualKeyCode, KeyQueryState>,
-    pressed_mouse_buttons: HashMap<MouseButton, KeyQueryState>,
+    pressed_keys: Vec<VirtualKeyCode>,
+    previous_pressed_keys: Vec<VirtualKeyCode>,
+    pressed_mouse_buttons: Vec<MouseButton>,
+    previous_pressed_mouse_buttons: Vec<MouseButton>,
     // defined to match glium
     mouse_position: Option<(i32, i32)>,
     previous_mouse_position: Option<(i32, i32)>,
+    axes: HashMap<i32, Axis>,
+    actions: HashMap<i32, Button>,
 }
 
 impl InputHandler {
     /// Creates a new input handler.
     pub fn new() -> InputHandler {
         InputHandler {
-            pressed_keys: HashMap::default(),
-            pressed_mouse_buttons: HashMap::default(),
+            pressed_keys: Vec::new(),
+            previous_pressed_keys: Vec::new(),
+            pressed_mouse_buttons: Vec::new(),
+            previous_pressed_mouse_buttons: Vec::new(),
             mouse_position: None,
             previous_mouse_position: None,
+            axes: HashMap::default(),
+            actions: HashMap::default(),
         }
     }
 
     /// Updates the input handler with new engine events.
     pub fn update(&mut self, events: &[WindowEvent]) {
+        // Before processing these events store the input states of the previous frame.
+        self.previous_mouse_position = self.mouse_position;
+        self.previous_pressed_keys = self.pressed_keys.clone();
+        self.previous_pressed_mouse_buttons = self.pressed_mouse_buttons.clone();
         for event in events {
             match event.payload {
                 Event::KeyboardInput(ElementState::Pressed, _, Some(key_code)) => {
-                    // The check for Entry::Vacant allows more accurate `key_once` calls,
-                    // I.e `key_once(key)` is queried after second `Pressed` event.
-                    if let Entry::Vacant(entry) = self.pressed_keys.entry(key_code) {
-                        entry.insert(KeyQueryState::NotQueried);
+                    if self.pressed_keys.iter().all(|&k| k != key_code) {
```

Comment:
It might be, although I think the performance of Vec might actually be better as iterating over <5 elements in contiguous memory is probably faster than any hashing function